### ISSUE TITLE
Calls to /assign before /useraliases cause unlinkable "alias" users

### ISSUE THREAD

**Issue by msandbothe** [CarnegieLearningWeb/UpGrade#69]
When /assign (and possible /mark) are called on a userId that does not exist, a new root user row is created for them in the DB. This can happen due to out of order calls or an unhandled failure in earlier init calls.

What solution should we provide to downgrade these users from root to alias of another user. It will need to handle possible inconsistencies in group exclusion.

**Comment by nirmalpatel** [comment]
So as I understand this, it looks like the alias user's assign call is coming in first, making the alias user a root user. And the root becomes the alias, and subsequent calls to the useralias API are then failing. If this is not the correct description, please let me know.

After some discussion, it seems like an init() call to explicitly initialize the root user might solve the problem. We can also stop auto-creating root users and only create root users through init() calls.

A vast number of things are linked with the root ID (assignments, metrics, many other things...) so if we downgrade them (in exchange for upgrading some other user), it will lead to a lot of cascading changes in the database. And so the scope of creating this functionality seems pretty big compared to the init() call solution.

**Comment by amurphy-cl** [comment]
@SritterCL do we want to redefine how we handle alias users before we proceed with this fix? e.g. create single UUID that isn’t a “root” but refers to a set of aliases, so UpGrade can accept multiple alias IDs that never get removed. If we can answer this, the next step will be to figure out the work involved in the solution (via PP input)

**Comment by SritterCL** [comment]
Yeah, assuming that this isn't adding a lot of extra work, I'd prefer a solution where we're not assuming one user ID has a different status from the others. I can imagine that with other applications (like if we do the LDI connector thing), students might have different IDs coming from different places with no reason to declare one of them the root.

**Comment by nirmalpatel** [comment]
@SritterCL so in a scenario where the portal call fails and subsequent calls succeed, and MATHia will first UpGrade with a ctcontextid. Then later, when the user comes back at the portal and the portal call succeeds, we will need both the user ID and the ctcontextid for the call otherwise the user ID may get a new random assignment...

**Comment by SritterCL** [comment]
Yeah - my assumption is that, at some point, UpGrade needs to get a message saying that ID1 and ID2 are for the same student and, at that point, UpGrade should merge those. The fact that MATHia considers ID1 to be a userID and ID2 to be a ctcontextID is irrelevant to UpGrade - they're just 2 IDs for the same user.

So, let's say portal fails and user ID1 gets assigned in Experiment 1. The user then logs off and back on. Now, portal succeeds and informs UpGrade that ID1 and ID2 are aliases. The user logs in with ID2 and gets the same assignment from Experiment 1.

The worst case scenario is if the portal fails multiple times. So portal fails, ID1 gets assigned in Experiment 1. Then user logs off and back on with ID2 and portal also fails. ID2 gets assigned in Experiment 1. Then the user logs in and portal succeeds, informing UpGrade that ID1 and ID2 are the same user. If we get lucky and the assignments are the same, the user records can be merged. If they're different, I think we just need to output an error message and treat ID1 and ID2 as different users (at least for that experiment - if it's possible to merge them for later experiments, that would be ideal).

**Comment by jerith666** [comment]
Steve, you weren't at the last status meeting, but I thought there we settled on an explicit `init` call for users, and then if UpGrade sees a `setAltIds` or `getAllExperimentConditions` request for a user it doesn't know about, it ignores it and/or returns default conditions for everything.  @amurphy-cl @msandbothe , sound right?

**Comment by amurphy-cl** [comment]
Yes, this sounds right, and iirc Playpower preferred this approach as well.

**Comment by nirmalpatel** [comment]
@SritterCL here are the scenarios that I worked out with one simplifying assumption: If UpGrade is up, all calls of one user are succeeding (this will be 99.9% I guess?)

With this assumption, we can look at these scenarios when UpGrade crashes and we have users online (0 means API fail, 1 means API success):

      Init            SetAltIDs               Assign              Mark
        0                   0                           0                       0      -  UpGrade is down and the user will get a reassignment when they visit the next time (initially they should get default)
        1                    0                           0                       0      -  User will get a reassignment because we were not able to mark that they got the default initially when the UpGrade was down
        1                     1                           0                       0      -  Same as above
        1                     1                           1                        0.     -  User won't get a reassignment, but they won't show up as enrolled in the experiment (as mark failed)

**Comment by nirmalpatel** [comment]
@jerith666 regarding the scenarios where earlier calls fail and later calls succeed, all of them will have init() failing and mark() succeeding, meaning that the user will get a default condition and we will know that they got the default condition next time they visit the same experiment site. So I think overall, the user will get excluded when their initial calls fail and later calls succeed.

**Comment by jerith666** [comment]
but the `mark()` call use the alias, not the root user ID.  Does that matter?

**Comment by nirmalpatel** [comment]
@jerith666 looking further, it looks like mark() creates a user if it does not exist. And if you are providing an alias in the mark(), then the actual marking happens with the root ID and not the alias. So if we disable user creation everywhere except in init(), then the mark() call with an alias or root ID will fail unless the init() call succeeded.

**Comment by amurphy-cl** [comment]
@SritterCL do we want to spec out further details about upgrade behavior if, e.g., upgrade is down in a student's session 1 and up in a session 2, in addition to other scenarios? Holding off on this until you weigh in.

**Comment by jerith666** [comment]
Yeah, I think having the `markExperimentPoint(userAlias)` fail if either of the preceeding calls -- `init()` or `setAltIds()` -- hasn't happened yet is fine.

**Comment by nirmalpatel** [comment]
We have made changes to the backend code and JS client library. Java client is remaining, hoping to get that done over the weekend...

**Comment by nirmalpatel** [comment]
Java client changes were made by Jaydip. Now the code reviews from PPL and CL are remaining.
